sparse/core.py

```diff
@@ -314,40 +314,58 @@ def __str__(self):
 
     __repr__ = __str__
 
-    def reduction(self, method, axis=None, keepdims=False, dtype=None):
+    @staticmethod
+    def _reduce(method, *args, **kwargs):
+        assert len(args) == 1
+
+        self = args[0]
+        if isinstance(self, scipy.sparse.spmatrix):
+            self = COO.from_scipy_sparse(self)
+
+        return self.reduce(method, **kwargs)
+
+    def reduce(self, method, axis=None, keepdims=False, **kwargs):
+        zero_reduce_result = method.reduce([_zero_of_dtype(self.dtype)], **kwargs)
+
+        if zero_reduce_result != _zero_of_dtype(np.dtype(zero_reduce_result)):
+            raise ValueError("Performing this reduction operation would produce "
+                             "a dense result: %s" % str(method))
```

Comment:
It seems to me that we might be ok with producing dense results with reductions, especially if we're losing most of our axes anyway.

Reply:
I think a `check` kwarg that defaults to `False` might be useful for later user configurability. This check is meant to be for a different reason, though, it checks that if we reduce across an all-zero axis, it shouldn't produce nonzeros. This might be useful in extremely sparse situations. If the overall result is dense because most axes are lost, this logic won't error out.

Reply:
Ah ok, that makes sense.  Thank you for the explanation.